File: Mika/tags/PKG-dap-1/core-vm/java/java/lang/reflect/Array.java
/**************************************************************************
* Copyright  (c) 2001 by Acunia N.V. All rights reserved.                 *
*                                                                         *
* This software is copyrighted by and is the sole property of Acunia N.V. *
* and its licensors, if any. All rights, title, ownership, or other       *
* interests in the software remain the property of Acunia N.V. and its    *
* licensors, if any.                                                      *
*                                                                         *
* This software may only be used in accordance with the corresponding     *
* license agreement. Any unauthorized use, duplication, transmission,     *
*  distribution or disclosure of this software is expressly forbidden.    *
*                                                                         *
* This Copyright notice may not be removed or modified without prior      *
* written consent of Acunia N.V.                                          *
*                                                                         *
* Acunia N.V. reserves the right to modify this software without notice.  *
*                                                                         *
*   Acunia N.V.                                                           *
*   Vanden Tymplestraat 35      info@acunia.com                           *
*   3000 Leuven                 http://www.acunia.com                     *
*   Belgium - EUROPE                                                      *
*                                                                         *
* Enhancements copyright (C) 2005 by Chris Gray, /k/ Embedded Java        *
* Solutions. All rights reserved.                                         *
*                                                                         *
**************************************************************************/

/*
** $Id: Array.java,v 1.2 2005/07/02 12:52:52 cvs Exp $
*/

package java.lang.reflect;

public final class Array {

  /*
  ** This class is a utility: it cannot be instantiated.
  */
  private Array(){}

  public static native Object newInstance(Class componentType, int length)
    throws NullPointerException, NegativeArraySizeException;
    
  public static native Object newInstance(Class componentType, int dimensions[])
    throws NullPointerException, IllegalArgumentException, NegativeArraySizeException;
  
  public static native int getLength(Object array)
    throws NullPointerException, IllegalArgumentException;
  
  public static native Object get(Object array, int index)
    throws NullPointerException, IllegalArgumentException, ArrayIndexOutOfBoundsException;

  public static boolean getBoolean(Object array, int index)
    throws NullPointerException, IllegalArgumentException, ArrayIndexOutOfBoundsException
  {
    boolean [] za;
    try {
      za = (boolean[])array;
    }
    catch (ClassCastException e) {
      throw new IllegalArgumentException(array + " is not an array of boolean");
    }
    return za[index];
  }

  public static byte getByte(Object array, int index)
    throws NullPointerException, IllegalArgumentException, ArrayIndexOutOfBoundsException
  {
    byte [] ba;
    try {
      ba = (byte[])array;
    }
    catch (ClassCastException e) {
      throw new IllegalArgumentException(array + " is not an array of byte");
    }
    return ba[index];
  }

  public static char getChar(Object array, int index)
    throws NullPointerException, IllegalArgumentException, ArrayIndexOutOfBoundsException
  {
    char [] ca;
    try {
      ca = (char[])array;
    }
    catch (ClassCastException e) {
      throw new IllegalArgumentException(array + " is not an array of char");
    }
    return ca[index];
  }

  public static short getShort(Object array, int index)
    throws NullPointerException, IllegalArgumentException, ArrayIndexOutOfBoundsException
  {
      Class arrayclass = array.getClass().getComponentType();
      if ( arrayclass == null) throw new IllegalArgumentException(array + " is not an Array");
      /* [CG 20050702] WAS:
      if (!arrayclass.isPrimitive()) throw new IllegalArgumentException(array + " is not an Array of primitive types");
      if (!arrayclass.toString().equals("byte") && !arrayclass.toString().equals("short"))
       	throw new IllegalArgumentException(array + " is not an Array of short or byte");

      Number nm;
      try {
          nm = (Number)Array.get(array , index);
          }
      catch (ClassCastException cce) { throw new IllegalArgumentException(array + " is not an Array of numeric primitive types");}
      return nm.shortValue();
      */

      if (arrayclass == byte.class) {
        return ((byte[])array)[index];
      }
      else if (arrayclass == short.class) {
        return ((short[])array)[index];
      }
      else {
       	throw new IllegalArgumentException(array + " is not an array of short or byte");
      }
  }

  public static int getInt(Object array, int index)
    throws NullPointerException, IllegalArgumentException, ArrayIndexOutOfBoundsException
  {
      Class arrayclass = array.getClass().getComponentType();
      if ( arrayclass == null) throw new IllegalArgumentException(array + " is not an Array");
    /* [CG 20050702] WAS:
      if (!arrayclass.isPrimitive()) throw new IllegalArgumentException(array + " is not an Array of primitive types");
      if (arrayclass.toString().equals("char")) { return (int)getChar(array,index);}
      if (!arrayclass.toString().equals("byte") && !arrayclass.toString().equals("short") && !arrayclass.toString().equals("int"))
      	throw new IllegalArgumentException("wrong primitive array type");
      Number nm;
      try {
          nm = (Number)Array.get(array , index);
          }
      catch (ClassCastException cce) { throw new IllegalArgumentException(array + " is not an Array of numeric primitive types");}
      return nm.intValue();
    */

      if (arrayclass == int.class) {
        return ((int[])array)[index];
      }
      else if (arrayclass == short.class) {
        return ((short[])array)[index];
      }
      else if (arrayclass == char.class) {
        return ((char[])array)[index];
      }
      else if (arrayclass == byte.class) {
        return ((byte[])array)[index];
      }
      else {
       	throw new IllegalArgumentException(array + " is not an array of int, char, short or byte");
      }
  }

  public static long getLong(Object array, int index)
    throws NullPointerException, IllegalArgumentException, ArrayIndexOutOfBoundsException
  {
      Class arrayclass = array.getClass().getComponentType();
      if ( arrayclass == null) throw new IllegalArgumentException(array + " is not an Array");
    /* [CG 20050702] WAS:
      if (!arrayclass.isPrimitive()) throw new IllegalArgumentException(array + " is not an Array of primitive types");
      if (arrayclass.toString().equals("char")) { return (long)getChar(array,index);}
      if (arrayclass.toString().equals("float") || arrayclass.toString().equals("double"))
      	throw new IllegalArgumentException("wrong primitive array type");
      Number nm;
      try {
          nm = (Number)Array.get(array , index);
          }
      catch (ClassCastException cce) { throw new IllegalArgumentException(array + " is not an Array of numeric primitive types");}
      return nm.longValue();
    */

      if (arrayclass == long.class) {
        return ((long[])array)[index];
      }
      else if (arrayclass == int.class) {
        return ((int[])array)[index];
      }
      else if (arrayclass == short.class) {
        return ((short[])array)[index];
      }
      else if (arrayclass == char.class) {
        return ((char[])array)[index];
      }
      else if (arrayclass == byte.class) {
        return ((byte[])array)[index];
      }
      else {
       	throw new IllegalArgumentException(array + " is not an array of int, char, short or byte");
      }
  }

  public static float getFloat(Object array, int index)
    throws NullPointerException, IllegalArgumentException, ArrayIndexOutOfBoundsException
  {
      Class arrayclass = array.getClass().getComponentType();
      if ( arrayclass == null) throw new IllegalArgumentException(array + " is not an Array");
    /* [CG 20050702] WAS:
      if (!arrayclass.isPrimitive()) throw new IllegalArgumentException(array + " is not an Array of primitive types");
      if (arrayclass.toString().equals("char")) { return (float)getChar(array,index);}
      if (arrayclass.toString().equals("double"))
      	throw new IllegalArgumentException("wrong primitive array type");
      Number nm;
      try {
          nm = (Number)Array.get(array , index);
          }
      catch (ClassCastException cce) { throw new IllegalArgumentException(array + " is not an Array of numeric primitive types");}
      return nm.floatValue();
    */

      if (arrayclass == float.class) {
        return ((float[])array)[index];
      }
      else if (arrayclass == long.class) {
        return ((long[])array)[index];
      }
      else if (arrayclass == int.class) {
        return ((int[])array)[index];
      }
      else if (arrayclass == short.class) {
        return ((short[])array)[index];
      }
      else if (arrayclass == char.class) {
        return ((char[])array)[index];
      }
      else if (arrayclass == byte.class) {
        return ((byte[])array)[index];
      }
      else {
       	throw new IllegalArgumentException(array + " is not an array of int, char, short or byte");
      }
  }

  public static double getDouble(Object array, int index)
    throws NullPointerException, IllegalArgumentException, ArrayIndexOutOfBoundsException
  {
      Class arrayclass = array.getClass().getComponentType();
      if ( arrayclass == null) throw new IllegalArgumentException(array + " is not an Array");
    /* [CG 20050702] WAS:
      if (!arrayclass.isPrimitive()) throw new IllegalArgumentException(array + " is not an Array of primitive types");
      if (arrayclass.toString().equals("char")) { return (double)getChar(array,index);}
      Number nm;
      try {
          nm = (Number)Array.get(array , index);
          }
      catch (ClassCastException cce) { throw new IllegalArgumentException(array + " is not an Array of numeric primitive types");}
      return nm.doubleValue();
    */

      // [CG 20050702] HAD ONCE BEEN:
    /*
    double [] da;
    try {
      da = (double[]) array;
    }
    catch (ClassCastException e) {
      throw new IllegalArgumentException();
    }
    return da[index]; 
    */

      // [CG 20050702] NOW:

      if (arrayclass == double.class) {
        return ((double[])array)[index];
      }
      else if (arrayclass == float.class) {
        return ((float[])array)[index];
      }
      else if (arrayclass == long.class) {
        return ((long[])array)[index];
      }
      else if (arrayclass == int.class) {
        return ((int[])array)[index];
      }
      else if (arrayclass == short.class) {
        return ((short[])array)[index];
      }
      else if (arrayclass == char.class) {
        return ((char[])array)[index];
      }
      else if (arrayclass == byte.class) {
        return ((byte[])array)[index];
      }
      else {
       	throw new IllegalArgumentException(array + " is not an array of int, char, short or byte");
      }
  }

  private static final Class Boolean_class = Boolean.class;
  private static final Class Byte_class = Byte.class;
  private static final Class Short_class = Short.class;
  private static final Class Character_class = Character.class;
  private static final Class Integer_class = Integer.class;
  private static final Class Long_class = Long.class;
  private static final Class Float_class = Float.class;
  private static final Class Double_class = Double.class;

  public static void set(Object array, int index, Object value)
    throws NullPointerException, IllegalArgumentException, ArrayIndexOutOfBoundsException 
  {
    Class arrayclass = array.getClass().getComponentType();
    if ( arrayclass == null) {
      throw new IllegalArgumentException(array + " is not an Array");
    }

    if (arrayclass.isPrimitive()) {
      Class value_class = value.getClass();
      if (value_class == Boolean_class && arrayclass == boolean.class) {
        ((boolean[])array)[index] = ((Boolean)value).booleanValue();
      }
      else if (value_class == Byte_class && arrayclass == byte.class) {
        ((byte[])array)[index] = ((Byte)value).byteValue();
      }
      else if (value_class == Byte_class && arrayclass == short.class) {
        ((short[])array)[index] = ((Byte)value).byteValue();
      }
      else if (value_class == Byte_class && arrayclass == int.class) {
        ((int[])array)[index] = ((Byte)value).byteValue();
      }
      else if (value_class == Byte_class && arrayclass == long.class) {
        ((long[])array)[index] = ((Byte)value).byteValue();
      }
      else if (value_class == Byte_class && arrayclass == float.class) {
        ((float[])array)[index] = ((Byte)value).byteValue();
      }
      else if (value_class == Byte_class && arrayclass == double.class) {
        ((double[])array)[index] = ((Byte)value).byteValue();
      }
      else if (value_class == Short_class && arrayclass == short.class) {
        ((short[])array)[index] = ((Short)value).shortValue();
      }
      else if (value_class == Short_class && arrayclass == int.class) {
        ((int[])array)[index] = ((Short)value).shortValue();
      }
      else if (value_class == Short_class && arrayclass == long.class) {
        ((long[])array)[index] = ((Short)value).shortValue();
      }
      else if (value_class == Short_class && arrayclass == float.class) {
        ((float[])array)[index] = ((Short)value).shortValue();
      }
      else if (value_class == Short_class && arrayclass == double.class) {
        ((double[])array)[index] = ((Short)value).shortValue();
      }
      else if (value_class == Character_class && arrayclass == char.class) {
        ((char[])array)[index] = ((Character)value).charValue();
      }
      else if (value_class == Character_class && arrayclass == int.class) {
        ((int[])array)[index] = ((Character)value).charValue();
      }
      else if (value_class == Character_class && arrayclass == long.class) {
        ((long[])array)[index] = ((Character)value).charValue();
      }
      else if (value_class == Character_class && arrayclass == float.class) {
        ((float[])array)[index] = ((Character)value).charValue();
      }
      else if (value_class == Character_class && arrayclass == double.class) {
        ((double[])array)[index] = ((Character)value).charValue();
      }
      else if (value_class == Integer_class && arrayclass == int.class) {
        ((int[])array)[index] = ((Integer)value).intValue();
      }
      else if (value_class == Integer_class && arrayclass == long.class) {
        ((long[])array)[index] = ((Integer)value).intValue();
      }
      else if (value_class == Integer_class && arrayclass == float.class) {
        ((float[])array)[index] = ((Integer)value).intValue();
      }
      else if (value_class == Integer_class && arrayclass == double.class) {
        ((double[])array)[index] = ((Integer)value).intValue();
      }
      else if (value_class == Long_class && arrayclass == long.class) {
        ((long[])array)[index] = ((Long)value).longValue();
      }
      else if (value_class == Long_class && arrayclass == float.class) {
        ((float[])array)[index] = ((Long)value).longValue();
      }
      else if (value_class == Long_class && arrayclass == double.class) {
        ((double[])array)[index] = ((Long)value).longValue();
      }
      else if (value_class == Float_class && arrayclass == float.class) {
        ((float[])array)[index] = ((Float)value).floatValue();
      }
      else if (value_class == Float_class && arrayclass == double.class) {
        ((double[])array)[index] = ((Float)value).floatValue();
      }
      else if (value_class == Double_class && arrayclass == double.class) {
        ((double[])array)[index] = ((Double)value).doubleValue();
      }
      else {
        throw new IllegalArgumentException("Cannot assign element of type " + value_class + " to array of " + arrayclass);
      }
    }
    else {
      try {
        ((Object[])array)[index] = value;
      }
      catch (ArrayStoreException ase) {
        throw new IllegalArgumentException("Cannot assign " + value + " to " + array);
      }
    }
  }

  public static void setBoolean(Object array, int index, boolean z)
    throws NullPointerException, IllegalArgumentException, ArrayIndexOutOfBoundsException {
    Class arrayclass = array.getClass().getComponentType();
    if (arrayclass == boolean.class) {
      ((boolean[])array)[index] = z;
    }
    else {
      throw new IllegalArgumentException(array + " is not an array of boolean");
    }
  }

  public static void setByte(Object array, int index, byte b)
    throws NullPointerException, IllegalArgumentException, ArrayIndexOutOfBoundsException {
    Class arrayclass = array.getClass().getComponentType();
    if (arrayclass == byte.class) {
      ((byte[])array)[index] = b;
    }
    else if (arrayclass == short.class) {
      ((short[])array)[index] = b;
    }
    else if (arrayclass == int.class) {
      ((int[])array)[index] = b;
    }
    else if (arrayclass == long.class) {
      ((long[])array)[index] = b;
    }
    else if (arrayclass == float.class) {
      ((float[])array)[index] = b;
    }
    else if (arrayclass == double.class) {
      ((double[])array)[index] = b;
    }
    else {
      throw new IllegalArgumentException(array + " is not an array of byte, short, int, long, float, or double");
    }
  }

  public static void setChar(Object array, int index, char c)
    throws NullPointerException, IllegalArgumentException, ArrayIndexOutOfBoundsException {
    Class arrayclass = array.getClass().getComponentType();
    if (arrayclass == char.class) {
      ((char[])array)[index] = c;
    }
    else if (arrayclass == int.class) {
      ((int[])array)[index] = c;
    }
    else if (arrayclass == long.class) {
      ((long[])array)[index] = c;
    }
    else if (arrayclass == float.class) {
      ((float[])array)[index] = c;
    }
    else if (arrayclass == double.class) {
      ((double[])array)[index] = c;
    }
    else {
      throw new IllegalArgumentException(array + " is not an array of char, int, long, float, or double");
    }
  }

  public static void setShort(Object array, int index, short s)
    throws NullPointerException, IllegalArgumentException, ArrayIndexOutOfBoundsException {
    Class arrayclass = array.getClass().getComponentType();
    if (arrayclass == short.class) {
      ((short[])array)[index] = s;
    }
    else if (arrayclass == int.class) {
      ((int[])array)[index] = s;
    }
    else if (arrayclass == long.class) {
      ((long[])array)[index] = s;
    }
    else if (arrayclass == float.class) {
      ((float[])array)[index] = s;
    }
    else if (arrayclass == double.class) {
      ((double[])array)[index] = s;
    }
    else {
      throw new IllegalArgumentException(array + " is not an array of short, int, long, float, or double");
    }
  }

  public static void setInt(Object array, int index, int i)
    throws NullPointerException, IllegalArgumentException, ArrayIndexOutOfBoundsException {
    Class arrayclass = array.getClass().getComponentType();
    if (arrayclass == int.class) {
      ((int[])array)[index] = i;
    }
    else if (arrayclass == long.class) {
      ((long[])array)[index] = i;
    }
    else if (arrayclass == float.class) {
      ((float[])array)[index] = i;
    }
    else if (arrayclass == double.class) {
      ((double[])array)[index] = i;
    }
    else {
      throw new IllegalArgumentException(array + " is not an array of int, long, float, or double");
    }

  }

  public static void setLong(Object array, int index, long l)
    throws NullPointerException, IllegalArgumentException, ArrayIndexOutOfBoundsException {
    Class arrayclass = array.getClass().getComponentType();
    if (arrayclass == long.class) {
      ((long[])array)[index] = l;
    }
    else if (arrayclass == float.class) {
      ((float[])array)[index] = l;
    }
    else if (arrayclass == double.class) {
      ((double[])array)[index] = l;
    }
    else {
      throw new IllegalArgumentException(array + " is not an array of long, float, or double");
    }
  }

  public static void setFloat(Object array, int index, float f)
  throws NullPointerException, IllegalArgumentException, ArrayIndexOutOfBoundsException {
    Class arrayclass = array.getClass().getComponentType();
    if (arrayclass == float.class) {
      ((float[])array)[index] = f;
    }
    else if (arrayclass == double.class) {
      ((double[])array)[index] = f;
    }
    else {
      throw new IllegalArgumentException(array + " is not an array of float or double");
    }
  }


  public static void setDouble(Object array, int index, double d)
    throws NullPointerException, IllegalArgumentException, ArrayIndexOutOfBoundsException {
    Class arrayclass = array.getClass().getComponentType();
    if (arrayclass == double.class) {
      ((double[])array)[index] = d;
    }
    else {
      throw new IllegalArgumentException(array + " is not an array of double");
    }
  }

}

File: Mika/trunk/core-vm/src/native/java/lang/reflect/Array.c
/**************************************************************************
* Copyright (c) 2001, 2002, 2003 by Acunia N.V. All rights reserved.      *
*                                                                         *
* This software is copyrighted by and is the sole property of Acunia N.V. *
* and its licensors, if any. All rights, title, ownership, or other       *
* interests in the software remain the property of Acunia N.V. and its    *
* licensors, if any.                                                      *
*                                                                         *
* This software may only be used in accordance with the corresponding     *
* license agreement. Any unauthorized use, duplication, transmission,     *
*  distribution or disclosure of this software is expressly forbidden.    *
*                                                                         *
* This Copyright notice may not be removed or modified without prior      *
* written consent of Acunia N.V.                                          *
*                                                                         *
* Acunia N.V. reserves the right to modify this software without notice.  *
*                                                                         *
*   Acunia N.V.                                                           *
*   Philips Site 5, box 3       info@acunia.com                           *
*   3000 Leuven                 http://www.acunia.com                     *
*   Belgium - EUROPE                                                      *
*                                                                         *
* Modifications copyright (c) 2004, 2006 by Chris Gray, /k/ Embedded Java *
* Solutions. All rights reserved.                                         *
*                                                                         *
**************************************************************************/

/*
** $Id: Array.c,v 1.5 2006/10/04 14:24:16 cvsroot Exp $
** 
** Implementation of the native methods for java/lang/reflect/Array
*/

#include <string.h>

#include "arrays.h"
#include "clazz.h"
#include "core-classes.h"
#include "descriptor.h"
#include "heap.h"
#include "loading.h"
#include "ts-mem.h"
#include "reflection.h"
#include "wstrings.h"
#include "threads.h"
#include "exception.h"

w_instance Array_static_get(JNIEnv *env, w_instance Class, w_instance Array, w_int idx) {

  w_thread thread = JNIEnv2w_thread(env);
  w_clazz clazz = instance2clazz(Array);
  w_clazz component_clazz;

  if (!Array) {
    throwException(thread, clazzNullPointerException, NULL);
    return NULL;
  }

  if (!clazz->dims) {
    throwException(thread, clazzIllegalArgumentException, "not an array class");
    return NULL;
  }

  if (idx < 0 || idx >= instance2Array_length(Array)) {
    throwException(thread, clazzArrayIndexOutOfBoundsException, NULL);
    return NULL;
  }

  component_clazz = clazz->previousDimension;

  if (clazzIsPrimitive(component_clazz)) {
    w_clazz    wrapper_clazz = NULL;
    w_instance wrapped;
    w_word     wrappee;
    w_double   vdouble;
    w_long     vlong;

    switch (component_clazz->type & 0x0f) {
      case VM_TYPE_BOOLEAN:
        wrapper_clazz = clazzBoolean;
        break;

      case VM_TYPE_CHAR:
        wrapper_clazz = clazzCharacter;
        break;

      case VM_TYPE_FLOAT:
        wrapper_clazz = clazzFloat;
        break;

      case VM_TYPE_DOUBLE:
        wrapper_clazz = clazzDouble;
        break;

      case VM_TYPE_BYTE:
        wrapper_clazz = clazzByte;
        break;

      case VM_TYPE_SHORT:
        wrapper_clazz = clazzShort;
        break;

      case VM_TYPE_INT:
        wrapper_clazz = clazzInteger;
        break;

      case VM_TYPE_LONG:
        wrapper_clazz = clazzLong;
        break;

      default:
        wabort(ABORT_WONKA, "Illegal primitive type 0x%02x\n");
    }

    wrapped = allocInstance(thread, wrapper_clazz);
    if (wrapped) {
      switch (component_clazz->type & 0x0f) {
        case VM_TYPE_BOOLEAN:
          {
            w_int byteidx = idx / 8;
            w_int bitidx = idx % 8;

            w_sbyte bytevalue = instance2Array_byte(Array)[byteidx];
            bytevalue = (bytevalue >> bitidx) & 1;
            wrappee = (w_word)bytevalue;
            setBooleanField(wrapped, F_Boolean_value, wrappee);
          }
          break;

        case VM_TYPE_CHAR:
          setCharacterField(wrapped, FIELD_OFFSET(F_Character_value), instance2Array_char(Array)[idx]);
          break;

        case VM_TYPE_FLOAT:
          setFloatField(wrapped, FIELD_OFFSET(F_Float_value), instance2Array_float(Array)[idx]);
          break;

        case VM_TYPE_DOUBLE:
          vdouble = instance2Array_double(Array)[idx];
          setDoubleField(wrapped, FIELD_OFFSET(F_Double_value), vdouble);
          break;

        case VM_TYPE_BYTE:
          setByteField(wrapped, FIELD_OFFSET(F_Byte_value), instance2Array_byte(Array)[idx]);
          break;

        case VM_TYPE_SHORT:
          setShortField(wrapped, FIELD_OFFSET(F_Short_value), instance2Array_short(Array)[idx]);
          break;

        case VM_TYPE_INT:
          setIntegerField(wrapped, FIELD_OFFSET(F_Integer_value), instance2Array_int(Array)[idx]);
          break;

        case VM_TYPE_LONG:
          vlong = instance2Array_long(Array)[idx];
          setLongField(wrapped, FIELD_OFFSET(F_Long_value), vlong);
          break;

        default:
          wabort(ABORT_WONKA, "Illegal primitive type 0x%02x\n");
      }

    }

    return wrapped;

  }
  else {
    return instance2Array_instance(Array)[idx];
  }

}

w_int Array_static_getLength(JNIEnv *env, w_instance Class, w_instance Array) {
  w_thread thread = JNIEnv2w_thread(env);

  if (!Array) {
  	throwException(thread, clazzNullPointerException, NULL);
  	return -1;
  }	
  if (!instance2clazz(Array)->dims){
    throwException(thread, clazzIllegalArgumentException, "not an array class");
    return -1;
  }	
  return instance2Array_length(Array);
  
}

/* WAS :
w_instance Array_static_newInstance_single(JNIEnv *env, w_instance Class, w_instance Component, w_int length) {

  w_thread thread = JNIEnv2w_thread(env);
  w_int dimensions;
  w_instance Array = NULL;
  w_clazz component_clazz = Class2clazz(Component);
  w_clazz array_clazz;

  if (length < 0) {
    throwException(thread, clazzNegativeArraySizeException, NULL);

    return NULL;

  }
  else if (!Component) {
    throwException(thread, clazzNullPointerException, NULL);

    return NULL;

  } else if(component_clazz == clazz_void) {
    throwException(thread, clazzIllegalArgumentException, NULL);
    return NULL;
  }

  woempa(9, "Creating instance of 1 dimension and component '%k'.\n", component_clazz);

  dimensions = length;
  array_clazz = getNextDimension(component_clazz);
  if (array_clazz) {
    if (!exceptionThrown(thread)) {
      mustBeLinked(array_clazz);
    }
    if (! exceptionThrown(thread)) {
      Array = allocArrayInstance_1d(thread, array_clazz, dimensions);
    }
  }

  return Array;
  
}
*/

w_instance Array_static_newInstance_single(JNIEnv *env, w_instance Class, w_instance Component, w_int length) {

  w_thread thread = JNIEnv2w_thread(env);
  w_instance Array = NULL;
  w_clazz component_clazz = Class2clazz(Component);
  w_clazz array_clazz;
  w_instance initiating_loader = clazz2loader(getCallingClazz(thread));
  

  if (length < 0) {
    throwException(thread, clazzNegativeArraySizeException, NULL);

    return NULL;

  }
  else if (!Component) {
    throwException(thread, clazzNullPointerException, NULL);

    return NULL;

  }
  else if(component_clazz == clazz_void) {
    throwException(thread, clazzIllegalArgumentException, NULL);
    return NULL;
  }

  woempa(9, "Creating instance of 1 dimension and component '%k'.\n", component_clazz);

  array_clazz = getNextDimension(component_clazz, initiating_loader);
  if (array_clazz) {
    if (!exceptionThrown(thread)) {
      mustBeLinked(array_clazz);
    }
    if (! exceptionThrown(thread)) {
      Array = allocArrayInstance_1d(thread, array_clazz, length);
    }
  }

  return Array;
  
}

w_instance Array_static_newInstance_multi(JNIEnv *env, w_instance Class, w_instance Component, w_instance Dimensions) {

  w_int * dimensions;
  w_int i;
  w_int ndims;
  w_int *lengths;
  w_instance Array = NULL;
  w_thread thread = JNIEnv2w_thread(env);
  w_clazz clazz;

  if (!Component || !Dimensions) {
    throwException(thread, clazzNullPointerException, NULL);
  }  
  else {
    woempa(9, "Creating instance of %d dimension(s) and component '%k'.\n", instance2Array_length(Dimensions), Class2clazz(Component));
    /*
    ** Check and copy our Dimensions elements...
    */

    ndims = instance2Array_length(Dimensions);
    if (ndims < 1 || ndims > 255) {
      throwException(thread, clazzIllegalArgumentException, "illegal number of dimensions");
    } 
    else {
      clazz = Class2clazz(Component);
    
      if (clazz == clazz_void) {
        throwException(thread, clazzIllegalArgumentException, NULL);
        return NULL;
      }


      /*
      ** Create an array of w_int for the length of each dimension...
      */

      dimensions = allocMem(ndims * sizeof(w_word));
      if (!dimensions) {
         return NULL;
      }
      lengths = instance2Array_int(Dimensions);
      for (i = 0; i < ndims; i++) {
        if (lengths[i] < 0) {
          woempa(9, "lengths[%d] = %d !!\n", i, lengths[i]);
          throwException(thread, clazzNegativeArraySizeException, NULL);
          break;
        }
        dimensions[i] = lengths[i];
      }

      /*
      ** Create now an array clazz from the dimensions and the component clazz
      */

      if (exceptionThrown(thread) == NULL) {
        w_instance initiating_loader = clazz2loader(getCallingClazz(thread));
  
        woempa(7, "Component class = %k\n", clazz);
        for (i = 0; i < ndims; i++) {
          clazz = getNextDimension(clazz, initiating_loader);
          if (exceptionThrown(thread)) {
            break;
          }
          woempa(7, "Resolving class %k\n", clazz);
          mustBeLinked(clazz);
          if (exceptionThrown(thread)) {
            break;
          }
          woempa(7, "Resolved class %k\n", clazz);
        }

        if (! exceptionThrown(thread)) {
          woempa(7, "Allocating %k\n", clazz);
          Array = allocArrayInstance(thread, clazz, ndims, dimensions);
        }
      }
      releaseMem(dimensions);
    }
  }
  
  return Array;
  
}

